snakefile.py

```diff
@@ -147,22 +159,34 @@ def tool(name):
             for name in generated:
                 print("  - {}".format(name))
 
+# function to pass read files to trim/filter/qc improvement
+def trim_reads_input(args):
+  sample = args[0]
+  return [os.path.join(READS_DIR, f) for f in lookup('name', sample, ['reads', 'reads2']) if f]
+
 
 rule prinseq:
-    input:
-        r1 = os.path.join(READS_DIR, "{sample}_R1.fastq"),
-        r2 = os.path.join(READS_DIR, "{sample}_R2.fastq")
+    input: trim_reads_input
     output:
-        r1 = os.path.join(TRIMMED_READS_DIR, "{sample}_trimmed_R1.fastq"),
-        r2 = os.path.join(TRIMMED_READS_DIR, "{sample}_trimmed_R2.fastq")
+        r1 = os.path.join(TRIMMED_READS_DIR, "{sample}_trimmed_R1.fastq.gz"),
+        r2 = os.path.join(TRIMMED_READS_DIR, "{sample}_trimmed_R2.fastq.gz")
     params:
         len_cutoff = int(READ_LENGTH * CUT_OFF),
         output = os.path.join(TRIMMED_READS_DIR, "{sample}"),
         tmp_r1 = os.path.join(TRIMMED_READS_DIR, "{sample}_1.fastq"),
-        tmp_r2 = os.path.join(TRIMMED_READS_DIR, "{sample}_2.fastq")
+        tmp_r2 = os.path.join(TRIMMED_READS_DIR, "{sample}_2.fastq"), 
+        read_dir = READS_DIR
     log: os.path.join(LOG_DIR, 'prinseq_{sample}.log')
-    shell: "{PRINSEQ_EXEC} -fastq {input.r1} -fastq2 {input.r2} -ns_max_n 4 -min_qual_mean 30 -trim_qual_left 30 -trim_qual_right 30 -trim_qual_window 10 -out_good {params.output} -out_bad null -min_len {params.len_cutoff} >> {log} 2>&1 && mv {params.tmp_r1} {output.r1} && mv {params.tmp_r2} {output.r2}"
-
+    shell: 
+      """ 
+        gunzip --keep {input[0]} {input[1]};
+        read1={input[0]} && read2={input[1]} && {PRINSEQ_EXEC} -fastq ${{read1%.*}} -fastq2 ${{read2%.*}} -ns_max_n 4\
+        -min_qual_mean 30 -trim_qual_left 30\
+        -trim_qual_right 30 -trim_qual_window 10 -out_good {params.output} -out_bad null -min_len {params.len_cutoff}\
+        >> {log} 2>&1; 
+        gzip {params.tmp_r1} && mv {params.tmp_r1}.gz {output.r1};
```

Comment:
You could use redirection instead of renaming the file.  On the other hand, if this is done in two steps to ensure that the file is created atomically please add a comment.

Reply:
I couldn't find how to do that with gzip. 
All the things I tried just resulted in empty files.
What would be the way for correct redirection then?